compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -150,18 +167,35 @@ impl Kind {
         matches!(self, Self::Numeric { .. })
     }
 
-    pub(crate) fn matches_opt(&self, other: Option<Self>) -> bool {
-        other.as_ref().map_or(true, |other_kind| self.unifies(other_kind))
-    }
-
     pub(crate) fn u32() -> Self {
         Self::Numeric(Box::new(Type::Integer(Signedness::Unsigned, IntegerBitSize::ThirtyTwo)))
     }
 
+    pub(crate) fn follow_bindings(&self) -> Self {
+        match self {
+            Self::Any => Self::Any,
+            Self::Normal => Self::Normal,
+            Self::Integer => Self::Integer,
+            Self::IntegerOrField => Self::IntegerOrField,
+            Self::Numeric(typ) => Self::Numeric(Box::new(typ.follow_bindings())),
+        }
+    }
+
     /// Unifies this kind with the other. Returns true on success
     pub(crate) fn unifies(&self, other: &Kind) -> bool {
         match (self, other) {
-            (Kind::Normal, Kind::Normal) => true,
+            // Kind::Any unifies with everything
+            (Kind::Any, _)
+            | (_, Kind::Any)
+
+            // Kind::Normal unifies with Kind::Integer and Kind::IntegerOrField
+            | (Kind::Normal, Kind::Integer | Kind::IntegerOrField)
+            | (Kind::Integer | Kind::IntegerOrField, Kind::Normal)
+
+            // Kind::Integer unifies with Kind::IntegerOrField
+            | (Kind::Integer | Kind::IntegerOrField, Kind::Integer | Kind::IntegerOrField) => true,
```

Comment:
```suggestion
            // Kind::Any unifies with everything
            (Kind::Any, _) | (_, Kind::Any) => true

            // Kind::Normal unifies with Kind::Integer and Kind::IntegerOrField
            (Kind::Normal, Kind::Integer | Kind::IntegerOrField)
            | (Kind::Integer | Kind::IntegerOrField, Kind::Normal) => true

            // Kind::Integer unifies with Kind::IntegerOrField
            (Kind::Integer | Kind::IntegerOrField, Kind::Integer | Kind::IntegerOrField) => true,
```

Nit but I think this is clearer where we just have separate branches for these cases. If clippy complains then we can silence it.